src/rdkafka_queue.h

```diff
@@ -1164,6 +1177,22 @@ rd_kafka_enq_once_disable(rd_kafka_enq_once_t *eonce) {
         return rko;
 }
 
+/**
+ * @brief Returns true if the queue can contain fetched messages.
+ *
+ * @locks rd_kafka_q_lock(rkq) if do_lock is set.
+ */
+static RD_INLINE RD_UNUSED rd_bool_t
+rd_kafka_q_can_contain_fetched_msgs(rd_kafka_q_t *rkq, rd_bool_t do_lock) {
```

Comment:
We can live without the lock if the `rkq` is a consumer queue because if the flag is already set, it is never unset. So if we first check the flag and it is already set, we don't need to lock the queue to get the value.

This benefits the performance as we are only concerned about the performance while consuming messages.

Reply:
We can't do this for correctness reasons even if the flag is only set once. 

Two reasons for this is. First is that that even if the poll is always guaranteed to be called by the user after setting of the flag is complete, it might still be on different threads, and without a memory barrier (like a lock) or unless flags is marked volatile, there is no guarantee that the two threads see the same value of the variable.

Second is that, while I'm reasonably certain that writes and reads from `int` are atomic in most cases, there isn't any requirement for them to be so.

Reply:
As discussed, First is not a problem. We need to check second one though.

Reply:
I was suggesting something like _[Double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking)_

Reply:
Resolving for now, will address later to see what we can trade-off wrt performance/correctness.